File: pkg/extensions/notify/notify.go
// Package notify sends native terminal notifications when the agent finishes.
//
// Supports multiple terminal protocols:
//   - OSC 777: Ghostty, iTerm2, WezTerm, rxvt-unicode
//   - OSC 99: Kitty
//
// When inside tmux, wraps sequences in DCS passthrough format.
//
// Import this package to enable the extension:
//
//	import _ "github.com/kfet/fir/pkg/extensions/notify"
package notify

import (
	"fmt"
	"io"
	"os"

	"github.com/kfet/fir/pkg/extension"
)

// output is the writer for notification escape sequences.
// Defaults to stderr to avoid interfering with the TUI's stdout rendering.
var output io.Writer = os.Stderr

// inTmux reports whether the process is running inside tmux.
var inTmux = func() bool {
	return os.Getenv("TMUX") != ""
}

// inKitty reports whether the process is running inside Kitty.
var inKitty = func() bool {
	return os.Getenv("KITTY_WINDOW_ID") != ""
}

func init() {
	extension.Register("notify", func(api extension.API) {
		api.On("agent_end", func(event *extension.Event, ctx extension.Context) (any, error) {
			notifyTerminal("Fi", "Ready for input")
			return nil, nil
		})
	})
}

// notifyTerminal sends a native terminal notification.
func notifyTerminal(title, body string) {
	if inKitty() {
		notifyOSC99(title, body)
	} else {
		notifyOSC777(title, body)
	}
}

// notifyOSC777 sends an OSC 777 notification (Ghostty, iTerm2, WezTerm, rxvt-unicode).
// When inside tmux, wraps the sequence in DCS passthrough format.
func notifyOSC777(title, body string) {
	if inTmux() {
		// DCS passthrough: wrap for tmux (every ESC becomes ESC ESC)
		// Format: ESC P tmux; ESC ESC ] 777;notify;title;body ESC ESC \ ESC \.
		fmt.Fprintf(output, "\x1bPtmux;\x1b\x1b]777;notify;%s;%s\x1b\x1b\\\x1b\\", title, body)
	} else {
		fmt.Fprintf(output, "\x1b]777;notify;%s;%s\x1b\\", title, body)
	}
}

// notifyOSC99 sends a Kitty OSC 99 notification.
// When inside tmux, wraps the sequences in DCS passthrough format.
func notifyOSC99(title, body string) {
	if inTmux() {
		fmt.Fprintf(output, "\x1bPtmux;\x1b\x1b]99;i=1:d=0;%s\x1b\x1b\\\x1b\\", title)
		fmt.Fprintf(output, "\x1bPtmux;\x1b\x1b]99;i=1:p=body;%s\x1b\x1b\\\x1b\\", body)
	} else {
		fmt.Fprintf(output, "\x1b]99;i=1:d=0;%s\x1b\\", title)
		fmt.Fprintf(output, "\x1b]99;i=1:p=body;%s\x1b\\", body)
	}
}

File: pkg/resources/builtin_extensions/notify.py
#!/usr/bin/env python3
# ---
# name: notify
# description: Send native terminal notifications when the agent finishes
# builtin: true
# modes: tui
# ---
"""Send native terminal notifications when the agent finishes.

Supports multiple terminal protocols:
  - OSC 777: Ghostty, iTerm2, WezTerm, rxvt-unicode
  - OSC 99: Kitty

When inside tmux, wraps sequences in DCS passthrough format.

This is a Python port of pkg/extensions/notify.
"""

import os
import re

import fir_ext


def _in_tmux():
    return os.environ.get("TMUX", "") != ""


def _group_id() -> str:
    """Stable per-tmux-session id for notification coalescing.

    Inside tmux, $TMUX is `socket,serverpid,sessionid` — globally unique per
    session. Sanitised for OSC 99 `i=` which wants `[A-Za-z0-9_-]`. Outside
    tmux, return "" so we fall back to the previous (non-grouped) behaviour.
    """
    tmux = os.environ.get("TMUX", "")
    if not tmux:
        return ""
    return re.sub(r"[^A-Za-z0-9_-]", "_", tmux)


def _in_kitty():
    return os.environ.get("KITTY_WINDOW_ID", "") != ""


def _write_to_tty(data: bytes) -> None:
    """Write bytes directly to /dev/tty, bypassing any captured pipes."""
    try:
        with open("/dev/tty", "wb", buffering=0) as tty:
            tty.write(data)
    except OSError:
        pass  # no controlling terminal (e.g. headless CI) — silently skip


def _notify_osc777(title: str, body: str):
    """OSC 777 notification (Ghostty, iTerm2, WezTerm, rxvt-unicode)."""
    if _in_tmux():
        _write_to_tty(f"\x1bPtmux;\x1b\x1b]777;notify;{title};{body}\x1b\x1b\\\x1b\\".encode())
    else:
        _write_to_tty(f"\x1b]777;notify;{title};{body}\x1b\\".encode())


def _notify_osc99(title: str, body: str):
    """Kitty OSC 99 notification."""
    ident = _group_id() or "1"
    if _in_tmux():
        _write_to_tty(f"\x1bPtmux;\x1b\x1b]99;i={ident}:d=0;{title}\x1b\x1b\\\x1b\\".encode())
        _write_to_tty(f"\x1bPtmux;\x1b\x1b]99;i={ident}:p=body;{body}\x1b\x1b\\\x1b\\".encode())
    else:
        _write_to_tty(f"\x1b]99;i={ident}:d=0;{title}\x1b\\".encode())
        _write_to_tty(f"\x1b]99;i={ident}:p=body;{body}\x1b\\".encode())


def notify_terminal(title: str, body: str):
    if _in_kitty():
        _notify_osc99(title, body)
    else:
        _notify_osc777(title, body)


_session_name = ""


@fir_ext.on("session_named")
def on_session_named(params, ctx):
    global _session_name
    _session_name = (params or {}).get("name", "")


@fir_ext.on("agent_end")
def on_agent_end(params, ctx):
    title = f"fir — {_session_name}" if _session_name else "fir"
    notify_terminal(title, "Ready for input")


fir_ext.run(name="notify")
